src/datasets/arrow_dataset.py

```diff
@@ -432,77 +434,110 @@ def __setstate__(self, state):
+                # Change path to relative path from inside the destination directory
+                data_file["filename"] = src.name
+
+            # Get state
+            state = self.__getstate__()
+            dataset_info = json.loads(state.pop("_info"))
+            assert state.get("_data") is None, "arrow table needs to be memory mapped"
+            assert state.get("_indices") is None, "arrow table needs to be memory mapped"
+            assert all(
+                len(h["transforms"]) == 0 for h in state.get("_inplace_history", [])
+            ), "in-place history needs to be empty"
+            # Serialize state
+            with fs.open(Path(dataset_path).joinpath("state.json").as_posix(), "w", encoding="utf-8") as state_file:
+                json.dump(state, state_file, indent=2, sort_keys=True)
+            with fs.open(
+                Path(dataset_path).joinpath("dataset_info.json").as_posix(), "w", encoding="utf-8"
+            ) as dataset_info_file:
+                json.dump(dataset_info, dataset_info_file, indent=2, sort_keys=True)
+            logger.info("Dataset saved in {}".format(dataset_path))
 
     @staticmethod
-    def load_from_disk(dataset_path: str) -> "Dataset":
-        """Load the dataset from a dataset directory
+    def load_from_disk(dataset_path: str, fs=None) -> "Dataset":
+        """
+        Loads a dataset that was previously saved using ``dataset.save_to_disk(dataset_path)`` from a filesystem using either :class:`datasets.filesystem.S3FileSystem` or ``fsspec.spec.AbstractFileSystem``.
 
         Args:
-            dataset_path (``str``): path of the dataset directory where the dataset will be loaded from
+            dataset_path (``str``): path (e.g. ``dataset/train``) or remote uri (e.g. ``s3://my-bucket/dataset/train``) of the dataset directory where the dataset will be loaded from
+            fs (Optional[:class:`datasets.filesystem.S3FileSystem`,``fsspec.spec.AbstractFileSystem``],  `optional`, defaults ``None``): instance of :class:`datasets.filesystem.S3FileSystem` or ``fsspec.spec.AbstractFileSystem`` used to download the files from remote filesystem.
+
+        Returns:
+            ``datasets.Dataset`` or ``datasets.DatasetDict``
+                if `dataset_path` is a path of a dataset directory: the dataset requested,
+                if `dataset_path` is a path of a dataset dict directory: a ``datasets.DatasetDict`` with each split.
         """
-        with open(os.path.join(dataset_path, "state.json"), "r", encoding="utf-8") as state_file:
+        # copies file from filesystem if it is remote filesystem to local filesystem and modifies dataset_path to temp directory containing local copies
+        if is_remote_filesystem(fs):
+            src_dataset_path = extract_path_from_uri(dataset_path)
+            tmp_dir = tempfile.TemporaryDirectory()
+            dataset_path = Path(tmp_dir.name).joinpath(src_dataset_path)
```

Comment:
```suggestion
            dataset_path = Path(tmp_dir).joinpath(src_dataset_path)
```
Here you want to use the tmp_dir no ?
Using tmp_dir.name may point to a relative directory